lib/sites-base.js

```diff
@@ -1,214 +1,326 @@
+          type: 'string'
+        },
+        {
+          name: 'productionHostname',
+          label: 'Production Hostname',
+          help: 'will add "www." if not present',
+          type: 'string'
+        }
+      ].concat(options.addFields || []);
 
-    body();
+      options.arrangeFields = [
+        {
+          name: 'urls',
+          label: 'URLs',
+          fields: [
+            'shortName',
+            'productionHostname',
+          ]
+        }
+      ].concat(options.arrangeFields || []);
+    } else {
+      options.addFields = [
+        {
+          type: 'array',
+          name: 'hostnamesArray',
+          label: 'Hostnames',
+          schema: [
+            {
+              type: 'string',
+              required: true,
+              name: 'hostname',
+              help: 'All valid hostnames for the site must be on this list, for instance both example.com and www.example.com'
+            }
+          ]
+        },
+        {
+          name: 'devBaseUrl',
+          label: 'Development Base URL',
+          help: 'like http://localhost:3000',
+          type: 'url'
+        },
+        {
+          name: 'stagingBaseUrl',
+          label: 'Staging Base URL',
+          help: 'like http://project.staging.org',
+          type: 'url'
+        },
+        {
+          name: 'prodBaseUrl',
+          label: 'Production Base URL',
+          help: 'like https://myproject.com',
+          type: 'url'
+        }
+      ].concat(options.addFields || []);
 
-    async function body() {
-      try {
-        await copyContent(req, piece, options);
-      } catch (e) {
-        return callback(e);
-      }
-      return callback(null);
+      options.arrangeFields = [
+        {
+          name: 'urls',
+          label: 'URLs',
+          fields: [
+            'hostnamesArray',
+            'devBaseUrl',
+            'stagingBaseUrl',
+            'prodBaseUrl'
+          ]
+        }
+      ].concat(options.arrangeFields || []);
     }
+  },
 
-    async function copyContent(req, piece, options) {
-
-      if (!piece.copyOfId) {
-        return;
+  construct: function(self, options) {
+    self.beforeSave = function(req, doc, options, callback) {
+      if ((!doc._id) && (!doc.copyOfId)) {
+        if (!doc.adminPassword) {
+          self.apos.notify(req, 'You must fill out the admin password field when creating a new site.', { type: 'error' });
+          return callback('adminPassword.required');
+        }
       }
+      // Don't let an admin password wind up in the db in cleartext
+      req.newAdminPassword = doc.adminPassword;
+      delete doc.adminPassword;
+  
+      if (self.options.baseUrlDomains) {
+        // new behavior
+        const shortname = doc.shortName || doc.slug;
+        doc.hostnames = []
 
-      let from, to;
+        _.forOwn(self.options.baseUrlDomains, function(value, key) {
+          doc[key+'BaseUrl'] = shortname + '.' + self.options.baseUrlDomains[key].toLowerCase().trim()
```

Comment:
Spaces here:

key+'BaseUrl'

I'm saying there should be spaces around the +. The lack of spaces around the brackets is fine.